src/index.js

```diff
@@ -183,22 +183,70 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function isFunctionDeclaration(node) {
+    return node && (t.is("FunctionDeclaration", node) || node.type === "NamedArrowDeclaration");
+  }
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+  function transformTails(path, allowLoops, getNewNode) {
+    path.resync();
 
-    let completionRecords = path.get("body").getCompletionRecords();
-    for (let targetPath of completionRecords) {
-      if (!targetPath.isExpressionStatement()) continue;
+    const tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (const tailPath of tailPaths) {
+      if (!tailPath) continue;
 
-      let arrPush = t.callExpression(
-        t.memberExpression(arrId, t.identifier("push")),
-        [targetPath.node.expression]
-      );
-      targetPath.replaceWith(arrPush);
+      if (tailPath.isExpressionStatement()) {
+        // TODO: add linting to discourage
+        if (tailPath.get("expression").isAssignmentExpression()) {
+          tailPath.insertAfter(getNewNode(tailPath.node.expression.left, tailPath));
+        } else {
+          tailPath.replaceWith(getNewNode(tailPath.node.expression, tailPath));
+        }
+      } else if (tailPath.isVariableDeclaration()) {
+        // TODO: handle declarations.length > 1
+        // TODO: add linting to discourage
+        tailPath.insertAfter(getNewNode(tailPath.node.declarations[0].id, tailPath));
+      } else if (isFunctionDeclaration(tailPath.node)) {
+        // Need to transform exprs to statements since this is block context.
+        const nextNode = getNewNode(tailPath.node.id, tailPath);
+        if (t.is("Expression", nextNode)) {
```

Comment:
@rattrayalex Not sure if you're missing anything, TBH. I was looking into the Babel code to introspect node types and I found that. I pulled on the thread and I thought if I called that method it would correctly understand the LightScript type aliases. (So I wouldn't need e.g.
```
    return node && (t.is("FunctionDeclaration", node) || node.type === "NamedArrowDeclaration");
```
to  find out if something is a function declaration.

However, it turns out it doesn't work, and I can't figure out why it doesn't work NamedArrowDeclaration definitely has FunctionDeclaration as an alias, and the code for `t.is` definitely checks the aliases. I was totally mystified.

I bet t.isExpression() would be fine here so I will amend.